docs/ota-client-guide/modules/ROOT/pages/virtual-secondaries.adoc

```diff
@@ -0,0 +1,55 @@
+= Use of Virtual Secondaries
+ifdef::env-github[]
+
+[NOTE]
+====
+We recommend that you link:https://docs.ota.here.com/ota-client/latest/{docname}.html[view this article in our documentation portal]. Not all of our articles render correctly in GitHub.
+====
+endif::[]
+
+
+Virtual secondary devices can be used as a way to update a configuration file on the primary device (where the primary device is a standard OSTree image running an aktualizr). You can do this in two parts:
+
+* First you need a virtual secondary configured that can be used to update a configuration file.
+* Afterwards you can integrate it into your Yocto build.
+
+
+
+== Creating a Virtual Secondary
+
+. You can create a virtual secondary to run locally by following the steps on how to xref:simulate-device-basic.adoc[simulate a device without building a disk image].
++
+You can use the provided `.deb` releases mentioned and the `virtualsec.json` file, which you can edit to suit your needs: such as removing one of the two entries and setting the paths and names to whatever you prefer.
+. Afterwards, you will be able to download the arbitrary files to the location specified in the "firmware_path" variable, by uploading the files to the https://connect.ota.here.com[OTA Connect Portal].
+. Now go to the *Software versions* tab and click *Add Software*. You should make sure the *ECU Types* field matches the value specified in the "ecu_hardware_id" variable of the `virtualsec.json` file.
++
+NOTE: "ecu_hardware_id" must be unique per each file type that you'll like to deliver, otherwise an update of one of them will rewrite the others.
+
+
+== Updating via Virtual Secondaries
+
+To issue the update perform the following steps:
+
+. Go to your *Devices* tab and find your primary device. (You can run 'aktualizr-info' locally to get the device name if you don't know it.)
+. Afterwards, select your virtual secondary and the file you uploaded to the *Software versions* tab.
++
+(For future reference, you can use the *Campaigns* tab to send updates to multiple devices simultaneously.)
+
+
+
+== Integrating Virtual Secondaries into Yocto
+
+You can use the files in the repository https://github.com/advancedtelematic/meta-updater/pull/611[here] as an example to get the virtual secondaries to work in Yocto. You can use the files directly or you can copy them to one of your layers and modify them according to your needs.

```

Comment:
@patrickvacek @xaviergully It doesn't look like a convenient way to add a custom update of a desired file by means of VirtualSecondary.

The current approach:

1. requires specifying non-important parameters, such as ecu_private_key, ecu_public_key, metadata_path, etc while a customer would be interested in just two of them in most cases, i.e. 'firmware_path' and 'ecu_hardware_id';

2. implies either modifying of the existing recipes or creating another layer and copy-pasting of the existing recipes.

Therefore, I'd suggest to do the following:

1. review and refactor the way Virtual Secondary are configured in order to minimize a number of parameters a customer needs to define;

2. review and refactor the virtual secondary recipes so a customer could specify just
     - a local.conf variable (e.g. VIRTUAL_SECONDARY_CONF) pointing to a customer's virtual secondary conf;
     - local.conf variables defining 'firmware_path' (and optionally ecu_hardware_id) and the virtual secondary recipes will do the rest of magic (e.g. auto-generate a virtual secondary configuration file).

Reply:
@mike-sul It's true, this isn't necessarily the easiest way to do this, and we can probably do better. For now, I think we should document how it works at present, and then we can come back later and iteratively improve what we have when we know more about what would be useful. Feel free to make a ticket if you'd like to get that conversation started.

@ammaboateng I think this is mostly fine as is, although I think we should change the link. It's currently pointing at a PR, which might be less helpful than pointing at the relevant files in the current repo instead.

Reply:
@patrickvacek can you please send me the link to the files that it should point to.

Reply:
Sure. The files changed in the PR are a pretty guide. The most important is `recipes-sota/config/aktualizr-virtualsec.bb`, and that links to the other two relevant files, so that's probably enough for now.

Reply:
Link: https://github.com/advancedtelematic/meta-updater/blob/master/recipes-sota/config/aktualizr-virtualsec.bb